core/src/main/java/jenkins/security/ApiTokenProperty.java

```diff
@@ -61,49 +73,90 @@
+    @SuppressFBWarnings(value = "MS_SHOULD_BE_FINAL", justification = "Accessible via System Groovy Scripts")
+    private static /* not final */ boolean SHOW_LEGACY_TOKEN_TO_ADMINS =
             SystemProperties.getBoolean(ApiTokenProperty.class.getName() + ".showTokenToAdmins");
     
+    /**
+     * If enabled, the users with {@link Jenkins#ADMINISTER} permissions can generate new tokens for
+     * other users. Normally a user can only generate tokens for himself.<p>
+     * Take care that only the creator of a token will have the plain value as it's only stored as an hash in the system.<p>
+     * Disabled by default due to the security reasons.
+     * It's the version of {@link #SHOW_LEGACY_TOKEN_TO_ADMINS} for the new API Token system (SECURITY-200).
+     *
+     * @since TODO
+     */
+    @SuppressFBWarnings(value = "MS_SHOULD_BE_FINAL", justification = "Accessible via System Groovy Scripts")
+    private static /* not final */ boolean ADMIN_CAN_GENERATE_NEW_TOKENS =
+            SystemProperties.getBoolean(ApiTokenProperty.class.getName() + ".adminCanGenerateNewTokens");
+
+    private volatile Secret apiToken;
+    private ApiTokenStore tokenStore;
     
     @DataBoundConstructor
     public ApiTokenProperty() {
-        _changeApiToken();
+        this.init();
     }
-
+    
+    public ApiTokenProperty readResolve() {
+        this.init();
+        return this;
+    }
+    
+    private void init() {
+        if (this.tokenStore == null) {
+            this.tokenStore = new ApiTokenStore();
+        }
+    }
+    
     /**
      * We don't let the external code set the API token,
      * but for the initial value of the token we need to compute the seed by ourselves.
      */
-    /*package*/ ApiTokenProperty(String seed) {
-        apiToken = Secret.fromString(seed);
+    /*package*/ ApiTokenProperty(@CheckForNull String seed) {
+        if(seed != null){
+            apiToken = Secret.fromString(seed);
+            tokenStore = new ApiTokenStore();
+            tokenStore.generateTokenFromLegacy(apiToken);
+        }
+        this.init();
     }
 
     /**
      * Gets the API token.
      * The method performs security checks since 1.638. Only the current user and SYSTEM may see it.
-     * Users with {@link Jenkins#ADMINISTER} may be allowed to do it using {@link #SHOW_TOKEN_TO_ADMINS}.
+     * Users with {@link Jenkins#ADMINISTER} may be allowed to do it using {@link #SHOW_LEGACY_TOKEN_TO_ADMINS}.
      * 
      * @return API Token. Never null, but may be {@link Messages#ApiTokenProperty_ChangeToken_TokenIsHidden()}
      *         if the user has no appropriate permissions.
      * @since 1.426, and since 1.638 the method performs security checks
      */
     @Nonnull
+    @SuppressFBWarnings("NP_NONNULL_RETURN_VIOLATION")
     public String getApiToken() {
-        return hasPermissionToSeeToken() ? getApiTokenInsecure() 
+        LOGGER.log(Level.WARNING, "Deprecated usage of getApiToken");
```

Comment:
This should probably log the stack trace (perhaps separately on FINE). Otherwise this information is useless.

It's also called when viewing the configure page of a user with a legacy token so I'm not sure a warning is appropriate here.